x/datapool/keeper/pool.go

```diff
@@ -553,3 +544,147 @@ func (k Keeper) increaseNumIssuedNFT(ctx sdk.Context, pool *types.Pool) {
 
 	k.SetPool(ctx, pool)
 }
+
+func (k Keeper) RedeemDataPass(ctx sdk.Context, redeemNFT types.MsgRedeemDataPass) (*types.DataPassRedeemReceipt, error) {
+	pool, err := k.GetPool(ctx, redeemNFT.PoolId)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if pool.GetRound() != redeemNFT.Round {
+		return nil, types.ErrRoundNotMatched
+	}
+
+	nftContractAcc, err := sdk.AccAddressFromBech32(pool.NftContractAddr)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress
+	}
+
+	moduleAddr := types.GetModuleAddress()
+	transferNFTMsg := types.NewTransferNFTMsg(moduleAddr.String(), strconv.FormatUint(redeemNFT.NftId, 10))
+
+	transferNFTMsgBz, err := json.Marshal(transferNFTMsg)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemerAcc, err := sdk.AccAddressFromBech32(redeemNFT.Redeemer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemTokenIds, err := k.GetRedeemerDataPassByAddr(ctx, pool.PoolId, redeemerAcc)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if !contains(redeemTokenIds, strconv.FormatUint(redeemNFT.NftId, 10)) {
+		return nil, types.ErrNotOwnedRedeemerNft
+	}
+
+	zeroFund := sdk.NewCoins(types.ZeroFund)
+
+	_, err = k.wasmKeeper.Execute(ctx, nftContractAcc, redeemerAcc, transferNFTMsgBz, zeroFund)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	nftRedeemReceipt := types.NewDataPassRedeemReceipt(redeemNFT.PoolId, redeemNFT.Round, redeemNFT.NftId, uint64(ctx.BlockHeight()), redeemNFT.Redeemer)
+
+	err = k.SetDataPassRedeemReceipt(ctx, *nftRedeemReceipt)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	return nftRedeemReceipt, nil
+}
+
+func (k Keeper) GetDataPassRedeemReceipt(ctx sdk.Context, poolID, round uint64, redeemer sdk.AccAddress) (types.DataPassRedeemReceipt, error) {
+	key := types.GetKeyPrefixNFTRedeemReceipt(poolID, round, redeemer)
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(key)
+
+	var receipt types.DataPassRedeemReceipt
+	err := k.cdc.UnmarshalBinaryLengthPrefixed(bz, &receipt)
+	if err != nil {
+		return types.DataPassRedeemReceipt{}, sdkerrors.Wrap(types.ErrGetDataPassRedeemReceipt, err.Error())
+	}
+
+	return receipt, nil
+}
+
+func (k Keeper) SetDataPassRedeemReceipt(ctx sdk.Context, redeemReceipt types.DataPassRedeemReceipt) error {
+	redeemer, err := sdk.AccAddressFromBech32(redeemReceipt.Redeemer)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, err.Error())
```

Comment:
Since this error is wrapped in the parent function, it would be good to just respond to the error.
```suggestion
		return err
```